src/main/scala/async/Listener.scala

```diff
@@ -1,73 +1,178 @@
 package gears.async
 
-import gears.async.Listener.Locked
-import gears.async.Listener.Failed
-import gears.async.Listener.SemiLock
 import scala.annotation.tailrec
 import gears.async.Async.Source
+import java.util.concurrent.locks.ReentrantLock
+
+/** A listener, representing an one-time value receiver of an [[Async.Source]].
+  *
+  * Most of the time listeners should involve only calling a receiver function,
+  * and can be created by [[Listener.apply]] or [[Listener.acceptingListener]].
+  *
+  * However, should the listener want to attempt synchronization, it has to
+  * expose some locking-related interfaces. See `lock` and `release`.
+  */
+trait Listener[-T]:
+  import Listener._
+  /** Complete the listener with the given item, from the given source.
+    * **If the listener exposes a [[Listener.ListenerLock]]**, it is required to acquire this lock completely
+    * (either through [[lockCompletely]] or through manual locking of every layer)
+    * before calling [[complete]]. This can also be done conveniently with [[completeNow]].
+    * For performance reasons, this condition is usually not checked and will end up
+    * causing unexpected behavior.
+    *
+    * The listener must automatically release the lock of itself and any underlying listeners,
+    * however this usually is done automatically by calling the inner listener's [[complete]]
+    * recursively.
+    */
+  def complete(data: T, source: Async.Source[T]): Unit
+
+  /** Represents the exposed API for synchronization on listeners at receiving time.
+    * If the listener does not have any form of synchronization, [[lock]] should be `null`.
+    */
+  val lock: Listener.ListenerLock | Null
+
+  /** Attempts to acquire all locks and then calling [[complete]] with the given item and source.
+    * If locking fails, [[releaseAll]] is automatically called.
+    */
+  def completeNow(data: T, source: Async.Source[T]): Boolean =
+    lockCompletely(source) match
+      case Locked =>
+        this.complete(data, source)
+        true
+      case Gone => false
+
+  /** Release the listener lock up to the given [[Listener.LockMarker]], if it exists. */
+  inline final def releaseLock(to: Listener.LockMarker): Unit = if lock != null then lock.releaseAll(to)
+
+  /** Attempts to completely lock the listener, if such a lock exists.
+    * Succeeds with [[Listener.Locked]] immediately if there is no [[Listener.ListenerLock]].
+    * If locking fails, [[releaseAll]] is automatically called.
+    */
+  def lockCompletely(source: Async.Source[T]): Locked.type | Gone.type =
+    this.lock match
+      case lock: Listener.ListenerLock => lock.lockAll(source)
+      case null => Locked
```

Comment:
Parallel to releaseLock
```suggestion
  inline final def lockCompletely(source: Async.Source[T]): Locked.type | Gone.type =
    if lock != null then lock.lockAll(source) else Locked
```